cpp/src/parquet/arrow/schema.h

```diff
@@ -163,24 +165,28 @@ struct PARQUET_EXPORT SchemaManifest {
     return it->second;
   }
 
-  bool GetFieldIndices(const std::vector<int>& column_indices, std::vector<int>* out) {
+  ::arrow::Result<std::vector<int>> GetFieldIndices(
+      const std::vector<int>& column_indices) {
     // Coalesce a list of schema field indices which are the roots of the
     // columns referred to by a list of column indices
     const schema::GroupNode* group = descr->group_node();
     std::unordered_set<int> already_added;
-    out->clear();
-    for (auto& column_idx : column_indices) {
+
+    std::vector<int> out;
+    for (int column_idx : column_indices) {
+      if (column_idx < 0 || column_idx >= descr->num_columns()) {
+        return ::arrow::Status::IndexError("Column index ", column_idx, " is not valid");
+      }
       auto field_node = descr->GetColumnRoot(column_idx);
       auto field_idx = group->FieldIndex(*field_node);
       if (field_idx < 0) {
```

Comment:
not your code, but is this really checking for equality to -1?

Reply:
You're right that the only negative value which `FieldIndex` returns is -1. IIUC this is may be overly defensive, though; is there a situation where we might fail to find a `Node` within a `GroupNode` when we'd just retrieved it from that same `GroupNode`?

Reply:
Worth noting: replacing this check with an assertion locally broke no tests in C++ or python. I'll leave it as an assertion for now and see what CI thinks

Reply:
https://github.com/apache/arrow/pull/7704#issuecomment-657119271
In this case the passed data is `column_indices`, which *are* checked. The field index is looked up from the group using a node pulled from that same group, so it seems redundant to check it again. I intended the assertion as a test for that intuition. If there is a way for `field_idx == -1` to occur then I agree that we should raise an exception rather than abort, but otherwise shouldn't the redundant check be removed?

Reply:
https://issues.apache.org/jira/browse/ARROW-9421